esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,259 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+        {
+            userDesc.setown(createUserDescriptor());
+            userDesc->set(userId.str(), context.queryPassword(), context.querySignature());
+        }
+
+        fileName.append("_").append(clusterName).append(".tmp");
+        VStringBuffer jobId("Create %s", fileName.str());
+        Owned<IFileDescriptor> fileDesc = createDFUDescriptor(context, fileName.str(), clusterName, jobId.str(), userId.str(), userDesc);
+        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fileDesc);
+        file->setAccessed();
+        file->setECL(recordDefinition);
+        file->queryAttributes().setPropBin("_rtlType", layoutBin.length(), layoutBin.toByteArray());
+        file->queryAttributes().setPropInt64("@recordCount", 0);
+        file->queryAttributes().setPropInt64("@size", 0);
+
+        getFileAccessBeforePublish(context, fileName.str(), jobId.str(), layoutBin, fileDesc, userDesc, accessType, requests, resp.updateAccessInfo());
+        resp.setName(fileName.str());
+        resp.setUserID(userId.str());
+        resp.setAccessType(requests.getAccessType());
+    }
+    catch (IException *e)
+    {
+        FORWARDEXCEPTION(context, e,  ECLWATCH_INTERNAL_ERROR);
+    }
+    return true;
+}
+
+bool CWsDfuEx::onDFUFilePublish(IEspContext &context, IEspDFUFilePublishRequest &req, IEspDFUFilePublishResponse &resp)
+{
+    try
+    {
+        SecAccessFlags accessType = translateToSecAccessFlags(req.getAccessType());
+        if (SecAccess_None == accessType)
+        {
+            context.setAuthStatus(AUTH_STATUS_NOACCESS);
+            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate - Permission denied.");
+        }
+        context.ensureFeatureAccess(FEATURE_URL, accessType, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate: Permission denied.");
+
+        const char *fileName = req.getName();
+        const char *cluster = req.getCluster();
+        const char *recordDefinition = req.getECLRecordDefinition();
+        if (isEmptyString(fileName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Name defined.");
+        if (isEmptyString(cluster))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Cluster defined.");
+        if (isEmptyString(recordDefinition))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No ECLRecordDefinition defined.");
+
+        MemoryBuffer layoutBin;
+        exportRecordDefinitionBinaryType(recordDefinition, layoutBin);
+
+        StringBuffer userId, oldName, newName;
+        unsigned len = strlen(fileName);
+        if ((len < (strlen(cluster) + 5)) || (strnicmp(fileName + len - 4, ".tmp", 4) != 0))
+        {//5: '_' and '.tmp'
+            newName.set(fileName);
+            oldName.set(fileName).append("_").append(cluster).append(".tmp");
+        }
+        else
+        {
+            oldName.set(fileName);
+            newName.append(len - 5 - strlen(cluster), fileName);
```

Comment:
newName.set()?

Reply:
I think that the append() is more efficient if only partial fileName is copied.